server/src/main/java/org/elasticsearch/action/bulk/BulkPrimaryExecutionContext.java

```diff
@@ -191,12 +193,38 @@ public void resetForMappingUpdateRetry() {
         resetForExecutionRetry();
     }
 
+    /**
+     * Don't bother the master node if the mapping update is a noop.
+     * This may happen if there was a concurrent mapping update that added the same field.
+     *
+     * @param mappingVersion the current mapping version. This is used to guard against infinite loops.
+     * @throws IllegalStateException if retried multiple times with the same mapping version, to guard against infinite loops.
+     */
+    public void resetForNoopMappingUpdateRetry(long mappingVersion) {
+        assert assertInvariants(ItemProcessingState.TRANSLATED);
+        if (noopMappingUpdateRetryForMappingVersion == mappingVersion) {
```

Comment:
I am not sure I follow why this infinite loop detection is necessary. I'd expect that since all dynamic fields from previous parse is now in the mappings, the retry would always succeed in one go?

The only case where I could see this not being true is updates, where a new version of the doc could appear and thus a new dynamic mapping update could potentially be required. That would be at odds with the detection here I think?

To some extent having some detection in assertions would be good, but it would have to not trigger for updates. And then be the same for other mapping update retries. I could accept not adding that now though.

Reply:
> You're referring to the situation when the retry_on_conflict parameter is set, right?

No, I am referring to a case where an update results in a mapping update and thus a retry afterwards. That could potentially be a noop mapping update (due to concurrency). The retried update could then result in new dynamic fields.

Reply:
I think it is safe towards such updates in the current form, since:

1. Any retry will at least use a mapping corresponding to the last mapping version, thus only adding fields missing there.
2. A noop result should thus not be possible.

I think #96235 does not change this, but the argument that we'll not run into it becomes more subtle.

A worry is that an update will fail with this error if it first adds a field f (noop detected) and then another field g (that is also a noop, towards same version). That seems impossible now, but hard to verify.

Reply:
About the potential discrepancy between what the mapper results in and what the document parser holds, I suggest to add testing to verify this as well as some assertion to ensure they are aligned at relevant times. That belongs to the other PR though (perhaps we can have the size tracking as a separate PR?).

Reply:
This is what https://github.com/elastic/elasticsearch/pull/102885 is about. It significantly simplifies the calculation of how a mapper counts against the field limit and aligns the implementations as much as possible (`Mapper.Builder#getTotalFieldsCount`, `Mapper#getTotalFieldsCount`, and `MappingLookup#getTotalFieldsCount`).

Reply:
I've also added an assertion here: https://github.com/elastic/elasticsearch/pull/102936/commits/befd7f6802ed2ec2eb1e05d16a9f64c0c46b971e